test/e2e/basic_provisioning_test.go

```diff
@@ -118,6 +123,143 @@ var _ = Describe("Provisioning", func() {
 			State:  metal3api.StateProvisioned,
 		}, e2eConfig.GetIntervals(specName, "wait-provisioned")...)
 
+		By("Retrieving the latest BMH object")
+		err = clusterProxy.GetClient().Get(ctx, client.ObjectKey{
+			Name:      bmh.Name,
+			Namespace: bmh.Namespace,
+		}, &bmh)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Adding the detached annotation")
+		helper, err = patch.NewHelper(&bmh, clusterProxy.GetClient())
+		Expect(err).NotTo(HaveOccurred())
+
+		// Add the detached annotation; "true" is used explicitly to clarify intent.
+		bmh.ObjectMeta.Annotations["baremetalhost.metal3.io/detached"] = "true"
+
+		Expect(helper.Patch(ctx, &bmh)).To(Succeed())
+
+		By("Saving the status to a JSON string")
+		savedStatus := bmh.Status
+		statusJSON, err := json.Marshal(savedStatus)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Deleting the BMH")
+		//Wait for 2 seconds to allow time to confirm annotation is set
+		time.Sleep(2 * time.Second)
+
+		err = clusterProxy.GetClient().Delete(ctx, &bmh)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting for the deleting state")
+		Eventually(func() (string, error) {
+			var currentBmh metal3api.BareMetalHost
+			err := clusterProxy.GetClient().Get(ctx, types.NamespacedName{Name: bmh.Name, Namespace: bmh.Namespace}, &currentBmh)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					// If the BMH is not found, we assume it has been successfully deleted.
+					return "deleting", nil
+				}
+				// Any other error should be returned.
+				return "", err
+			}
+
+			currentStatus := currentBmh.Status.Provisioning.State
+
+			// If the state is 'deleting', we return this state to finish the polling.
+			if currentStatus == "deleting" {
+				return "deleting", nil
+			}
+
+			// If the state is empty or 'provisioned', we continue polling.
+			if currentStatus == "" || currentStatus == "provisioned" {
+				return "", nil
+			}
```

Comment:
I have some reservations about the suggestion you've made. If we return the `currentStatus` when it's in the `provisioned` state (which is the state before it transitions to the state we're monitoring for) or when it's empty `""`, our test is likely to fail. These conditions, `provisioned` and `""`, represent transitional states leading up to the desired state we need to observe. Therefore, returning the `currentStatus` in these instances isn't suitable. We should continue polling in these cases rather than returning the `currentState`. That's my understanding of how it should work.

Reply:
Hmm maybe I am missing something. This is how I understand the code now:

```go
if currentStatus == "deleting" {
  // This is the same as returning currentStatus
  return "deleting", nil
}
if currentStatus == "" || currentStatus == "provisioned" {
  // This is the same as currentStatus when "". Returning "provisioned" would be different
  // but should not break the Eventually since we are waiting for "deleting"/"deleted".
  return "", nil
}
```

Based on this I would suggest changing it to 
```suggestion
			// If the state is 'deleting', 'provisioned' or '', we return this state.
			// Other states are unexpected
			if currentStatus == "deleting" || currentStatus == "" || currentStatus == "provisioned" {
				return currentStatus, nil
			}
```
It will only reach `StopTrying` if the status is something else. The `Eventually` is waiting for `"deleting"`, so it will continue trying if the status is something different and we return that without reaching `StopTrying`.